Signal/src/contact/OWSContactsManager.m

```diff
@@ -215,14 +215,10 @@ - (void)updateCachedDisplayNames
 {
     OWSAssert([NSThread isMainThread]);
 
-    // Preserve any existing values, so that contacts that have been removed
-    // from system contacts still show up properly in the app.
-    NSMutableDictionary<NSString *, NSString *> *cachedAccountNameMap
-        = (self.cachedAccountNameMap ? [self.cachedAccountNameMap mutableCopy] : [NSMutableDictionary new]);
-    NSMutableDictionary<NSString *, NSString *> *cachedFirstNameMap
-        = (self.cachedFirstNameMap ? [self.cachedFirstNameMap mutableCopy] : [NSMutableDictionary new]);
-    NSMutableDictionary<NSString *, NSString *> *cachedLastNameMap
-        = (self.cachedLastNameMap ? [self.cachedLastNameMap mutableCopy] : [NSMutableDictionary new]);
```

Comment:
For testing I often want to remove someone from my contacts - and the previous caching logic prevents that. I agree that's weird. So 👍 for addressing this.

However, I believe your change introduces a regression where no-longer-registered users will show up as phone numbers instead of using their name from the address book.

I have a feeling that the deeper problem is that we should be populating these caches not by iterating SignalAccounts, but rather by iterating Contacts - since we use them to display Contact's display names (some of whom do not have a SignalAccount).

The pickle is that disambiguating multi-number contacts is done on the SignalAccount. Maybe **that** disambiguation could be moved to the Contact?

Seems like a non-trivial refactor. FWIW I have a slight preference for the former behavior, than to what's proposed here, even if it makes testing harder, because I think it will seem less broken for more users.